### PR TITLE
fix(spec): gate response schema by destination provenance to the writer (#170)

### LINKED ISSUE #170: Response schema attributed to a value encoded to a non-ResponseWriter `io.Writer` (false positive)
Verified on current `main`. A value encoded to some **other** `io.Writer` (not the response writer) is wrongly emitted as the operation's response schema — there's no write-destination gating. GAP §7.3 "Write-destination gating".

### Reproduce
```go
func leak(w http.ResponseWriter, r *http.Request) {
    var buf bytes.Buffer
    json.NewEncoder(&buf).Encode(Secret{})  // encoded to buf, NOT to w
    w.Write([]byte("ok"))
}
type Secret struct { Token string `json:"token"` }
```

### Actual
```yaml
/leak:
  post:
    responses:
      default:
        content: { application/json: { schema: { $ref: '.../Secret' } } }
```
`Secret` never reaches the response writer, yet it's emitted as the response.

### Expected
A write should only become a response when its destination provably traces to the `http.ResponseWriter`.

### PR DESCRIPTION
**Pull request by ehabterra**
## Summary

Fixes #170. A value encoded to a non-response `io.Writer` — a `bytes.Buffer`, a hash, `io.Discard` — was wrongly emitted as the operation's response.

The key reframing (thanks to review feedback): **the response writer's concrete identity is already in the code** — it is the handler's `w http.ResponseWriter` parameter. So this is a **provenance question**, not a type-relationship computation (no VTA, no `go/types.Implements`, no denylist). An encode is a response only when its destination traces — through **parameters, assignments, and struct construction** — back to a writer-typed value. This is the exact mirror of the request-body source gating (#153, which traces a decoder's source to `r.Body`).

## Design

- `responseDestResolver` is the structural twin of `bodySourceResolver`.
- The gate lives in the **per-route `ExtractResponse`**, *not* the per-edge-memoized `MatchNode` — a helper shared by several routes resolves its destination to the actual argument at each route's call site (`encodeTo(w,v)` → `w`, `encodeTo(&buf,v)` → `&buf`).
- Config is framework-agnostic (`ResponseContext.WriterTypeRegexes` / `WriterCompatibleTypeRegexes`), populated for the net/http family (net/http, chi, mux); gin/echo/fiber stay permissive until their writer types are declared.

## Extensively tested — assignment & parameter matrix

| Shape | Result |
|---|---|
| `NewEncoder(w)` | **keep** ✓ |
| `encodeTo(w, v)` (io.Writer helper param) | **keep** ✓ |
| `dst := w` | **keep** ✓ |
| `a := w; b := a` (multi-assign) | **keep** ✓ |
| `var d io.Writer = …; d = w` (reassign to writer) | **keep** ✓ |
| `&loggingWriter{w}` / `{ResponseWriter: w}` (wrapper) | **keep** ✓ |
| `var buf bytes.Buffer; NewEncoder(&buf)` | **drop** ✓ |
| `encodeTo(&buf, v)` | **drop** ✓ |
| `buf := &bytes.Buffer{}` | **drop** ✓ |
| `sha256.New()` (hash) | **drop** ✓ |
| `io.Discard` | **drop** ✓ |
| `var d io.Writer = &bytes.Buffer{}` | **drop** ✓ |
| `var d io.Writer = w; d = &bytes.Buffer{}` (reassign to sink) | **drop** ✓ |
| `NewEncoder(makeBuffer())` (constructor-returned) | **drop** ✓ |

All 14 shapes classify correctly; unresolved/interface destinations stay permissive ("honest over wrong").

## Issue filed during testing

Per the "fix the root cause or file an issue, never a workaround" rule, extensive testing surfaced a **pre-existing** value-tracing gap (unrelated to this gate): a value forwarded through **2+** `any`-typed helper hops loses its concrete type (`{type: object}`). Reproduced identically on `main`. Filed as **#180**. This gate correctly *keeps* such responses (it only drops provable non-writers).

## Tests

- `testdata/response_writer_provenance/` — 4 keep + 4 drop shapes.
- `generator/testdata_response_writer_provenance_test.go` — structural: KEEP routes ref `User`, DROP routes carry no `$ref`, `Secret` never appears.
- `internal/spec/response_dest_test.go` — unit tests for direct types, address-of, latest-wins assignments, wrapper construction, and cyclic-assignment termination.

## Verification

Original #170 reproduction (`json.NewEncoder(&buf).Encode(Secret{})`) no longer leaks. Full `go test ./...` passes with **zero output drift**; `gofmt`, `go vet`, `golangci-lint` all clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved API response detection to distinguish actual HTTP responses from writes to unrelated destinations.
  * Prevented leaked or indirect data from being incorrectly included in generated response schemas.
  * Added support for tracing response destinations through assignments, wrappers, pointers, and receiver-based encoders.

* **Tests**
  * Added coverage for response provenance, writer detection, assignment handling, and edge cases across supported HTTP configurations.

### CODE REVIEW

File: internal/spec/response_dest.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+	if !r.Enabled() {
+		return false
+	}
+	if r.reachesWriter(arg, edge, make(map[string]bool, 4)) {
+		return false // provenance reaches the response writer — it's the response
+	}
+	t := r.leafType(arg, edge, make(map[string]bool, 4))
+	if t == "" || matchAny(r.compatibleREs, t) || matchAny(r.writerTypeREs, t) {
+		return false // unresolved, a writer-compatible interface, or a writer — keep
+	}
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/spec/response_dest.go:83]: Done in 04cfc8a — good catch, this was the right refinement. A writer *type* is now treated as insufficient proof:

- `reachesWriterIdent` follows local assignments FIRST; a local (an ident with an assignment in scope) is judged by its assigned provenance, and only an **unassigned** ident — the handler/helper parameter — is accepted on its writer type. So `rec := httptest.NewRecorder()` traces to the constructor (no `w`) and drops.
- `httptest.ResponseRecorder` and `net/http.response` removed from `WriterTypeRegexes` (independently-constructible concretes are not the handler's `w`).
- `ShouldDrop` no longer keeps a value merely because its leaf type is writer-shaped.

Verified: `json.NewEncoder(httptest.NewRecorder()).Encode(Secret{})` (direct and via a `rec :=` var) now drops, while direct `w` / io.Writer helper / `dst := w` / `&loggingWriter{w}` still keep. Added a `/leak-recorder` fixture route and flipped the unit-test expectation.

**Comment by ehabterra** [comment]
### Update (04cfc8a): CodeRabbit fix + symmetric request-side gating

**CodeRabbit — resolved.** A writer *type* is no longer accepted as provenance. `reachesWriterIdent` judges locals by their assigned provenance and only accepts an unassigned ident (the handler/helper parameter) on its writer type; `httptest.ResponseRecorder` / `net/http.response` are removed from `WriterTypeRegexes`; `ShouldDrop` no longer keeps writer-shaped leaves. So `json.NewEncoder(httptest.NewRecorder()).Encode(...)` drops (direct + via var). Added a `/leak-recorder` fixture route.

**Request side — same treatment (read-side mirror).** Moved the `RequireRequestSource` gate (#153) from the per-edge-memoized `RequestPatternMatcher.MatchNode` to the per-route `ExtractRequest`, resolving the decoder source through the call graph per route. This fixes:
- **false negative:** a decoder wrapped in `func decodeFrom(src io.Reader, v)` called with `r.Body` is now recognised;
- **false positive:** the same helper called with a `bytes.Buffer` / `strings.Reader` no longer leaks as a body (the shared-node problem a per-edge gate had);
- and `bodySourceResolver.checkIdent` now follows the handler function's local assignments (`src := r.Body`), matching the response side.

New `request_body_source_provenance` fixture (direct/helper/assign keep; buffer/string drop) + structural test. Full suite passes with **zero drift**; fmt/vet/lint clean.
